manager/allocator/allocator_new.go

```diff
@@ -0,0 +1,993 @@
+					"an overlay network was removed, deallocating it from nodes",
+				)
+				if err := a.store.Update(func(tx store.Tx) error {
+					storeDone := metrics.StartTimer(storeLockHeld.WithValues("network", "deallocate", "write"))
+					defer storeDone()
+
+					// see NOTE(dperny): FindByAll
+					nodes, _ := store.FindNodes(tx, store.All)
+					for _, node := range nodes {
+						a.pendingNodes[node.ID] = struct{}{}
+						nwids := map[string]struct{}{}
+						for _, attachment := range node.Attachments {
+							if attachment.Network.ID != ev.Network.ID {
+								nwids[attachment.Network.ID] = struct{}{}
+							}
+						}
+						allocDone := metrics.StartTimer(allocatorActions.WithValues("network", "allocate"))
+						err := a.network.AllocateNode(node, nwids)
+						allocDone()
+						if err != nil {
+							if !errors.IsErrAlreadyAllocated(err) {
+								// it shouldn't be possible for this
+								// reallocation to fail. if it does fail, our
+								// only option is to return an error and log
+								// it.
+								return err
+							}
+						}
+						if err := store.UpdateNode(tx, node); err != nil {
+							return err
+						}
+					}
+					return nil
+				}); err != nil {
+					log.G(ctx).WithError(err).Error("deallocating nodes for an overlay network failed")
+				}
+			}
+			allocDone := metrics.StartTimer(allocatorActions.WithValues("network", "deallocate"))
+			err := a.network.DeallocateNetwork(ev.Network)
+			allocDone()
+			if err != nil {
+				log.G(ctx).WithError(err).WithField("network.id", ev.Network.ID).Error("error deallocating network")
+			}
+			delete(a.pendingNetworks, ev.Network.ID)
+		}
+	// case api.EventCreateService, api.EventUpdateService:
+	//     the case for create and update services is not needed.
+	//     we will lazy-allocate services. the reason for this is
+	//     to guarantee that the latest version of the service is
+	//     always allocated, and the task is always using the
+	//     latest version of the service for allocation.
+	case api.EventDeleteService:
+		if ev.Service != nil {
+			if _, ok := a.deletedObjects[ev.Service.ID]; ok {
+				delete(a.deletedObjects, ev.Service.ID)
+				return
+			}
+			allocDone := metrics.StartTimer(allocatorActions.WithValues("service", "deallocate"))
+			err := a.network.DeallocateService(ev.Service)
+			allocDone()
+			if err != nil {
+				log.G(ctx).WithField("service.id", ev.Service.ID).WithError(err).Error("error deallocating service")
+			}
+		}
+	case api.EventCreateTask, api.EventUpdateTask:
+		var t *api.Task
+		if e, ok := ev.(api.EventCreateTask); ok {
+			t = e.Task
+		} else {
+			t = ev.(api.EventUpdateTask).Task
+		}
+		if t != nil {
+			// updating a task may mean the task has entered a
+			// terminal state. if it has, we will free its network
+			// resources just as if we had deleted it.
+			if t.Status.State >= api.TaskStateCompleted {
+				if _, ok := a.deletedObjects[t.ID]; ok {
+					delete(a.deletedObjects, t.ID)
```

Comment:
What happens in the case that a task is stopped and then deleted?  

If a task was added to `pendingTasks`, and we allocate the resources for it, but it was actually stopped and then deleted before we managed to commit, we would:

1.  Deallocate the task immediately, because we'd already allocated some resources for it and the object has no record of this, and append the object to `deletedObjects`.
2. We'd get the `api.EventUpdateTask`, notice that the task state >= `api.TaskStateCompleted`, and notice that it's in the `deletedObjects` map, and remove it from the `deletedObjects` map.  But the task remains in the store with the previous allocation.
3. We'd get the `api.EventDeleteTask`, which would then notice that the task ID is `not` in the `deletedObjects` map, and attempt to deallocate the task again.

Is this ok because the if it went from needing an allocation update to >= `api.TaskStateCompleted`, that means that the task in the store would have no objects allocated to it, and hence it's fine to deallocate the task again?  If so, do we need to add it to the `deletedObjects` map at all?

When we deallocate a task due to it being completed, we don't seem to update the task in the store to remove any record of resources allocated to it.

Similarly, even if it hadn't already been deallocated, what happens if it goes from complete to then being deleted?  Since we receive 2 events, are we going to deallocate twice?

Reply:
Yes, we need to check in deletion if a task has entered a terminal state. if it has already entered a terminal state, then it will have been deallocated when it did so, and does not need to be deallocated on deletion. I've fixed this. Great catch, that was a landmine.